File: rust/src/board.rs
//! # Board
//! Defines the game board.
//!
//! Akseli Lukkarila
//! 2019-2025

use std::cmp::Ordering;
use std::collections::HashSet;
use std::fmt;

use colored::{ColoredString, Colorize};

use crate::models::{Direction, Disk, Move, Square, Step};

const UP: isize = 1;
const DOWN: isize = -1;
const LEFT: isize = -1;
const RIGHT: isize = 1;
const STILL: isize = 0;

/// All possible step directions for a square on the board.
static STEP_DIRECTIONS: [Step; 8] = [
    Step { x: UP, y: LEFT },
    Step { x: UP, y: STILL },
    Step { x: UP, y: RIGHT },
    Step { x: STILL, y: LEFT },
    Step { x: STILL, y: RIGHT },
    Step { x: DOWN, y: LEFT },
    Step { x: DOWN, y: STILL },
    Step { x: DOWN, y: RIGHT },
];

/// Handles game board state and logic.
pub struct Board {
    board: Vec<Disk>,
    empty_squares: HashSet<Square>,
    indices: Box<[usize]>,
    size: usize,
}

impl Board {
    pub fn new(size: usize) -> Self {
        let board = Self::initialize_board(size);
        // Index list (0...size) to avoid repeating same range in loops.
        // Not really needed in Rust but kept here to more closely match other implementations...
        // Could just use `0..self.size` directly.
        let indices = (0..size).collect::<Vec<_>>().into_boxed_slice();

        // Keep track of empty squares on board to avoid checking already filled positions.
        let empty_squares = Self::initialize_empty_squares(size, &board);

        Self {
            board,
            empty_squares,
            indices,
            size,
        }
    }

    /// Return true if board contains empty squares.
    pub fn can_play(&self) -> bool {
        !self.empty_squares.is_empty()
    }

    /// Update board for given disk placement.
    pub fn place_disk(&mut self, player_move: &Move) {
        let start = player_move.square;
        if self
            .get_square(&start)
            .unwrap_or_else(|| panic!("Invalid coordinates: {start}"))
            != Disk::Empty
        {
            panic!("Trying to place disk to an occupied square: {start}!");
        }
        self.set_square(&start, player_move.disk);
        self.empty_squares.remove(&start);
        for square in &player_move.affected_squares() {
            self.set_square(square, player_move.disk);
        }
    }

    /// Returns a list of possible moves for the given player.
    pub fn possible_moves(&self, disk: Disk) -> Vec<Move> {
        let mut moves: Vec<Move> = Vec::new();
        let opposing_disk = disk.opponent();
        for square in &self.empty_squares {
            let mut value: usize = 0;
            let mut directions: Vec<Direction> = Vec::new();
            for step in &STEP_DIRECTIONS {
                let mut pos = *square + *step;
                // Next square in this direction needs to be the opposing disk
                if self.get_square(&pos).unwrap_or(Disk::Empty) != opposing_disk {
                    continue;
                }
                let mut num_steps: usize = 0;
                // Keep stepping over opponents disks
                while self.get_square(&pos).unwrap_or(Disk::Empty) == opposing_disk {
                    num_steps += 1;
                    pos += *step;
                }
                // Valid move only if a line of opposing disks ends in own disk
                if self.get_square(&pos).unwrap_or(Disk::Empty) == disk {
                    directions.push(Direction::new(*step, num_steps));
                    value += num_steps;
                }
            }
            if value > 0 {
                moves.push(Move {
                    square: *square,
                    disk,
                    value,
                    directions,
                });
            }
        }
        if !moves.is_empty() {
            moves.sort();
        }
        moves
    }

    /// Print board with available move coordinates and the resulting points gained.
    pub fn print_possible_moves(&self, moves: &Vec<Move>) {
        println!(
            "{}",
            format!("  Possible moves ({}):", moves.len()).yellow()
        );
        // Convert board from Disk enums to strings
        let mut formatted_board: Vec<ColoredString> = self
            .board
            .iter()
            .map(|&d| d.board_char_with_color())
            .collect();

        // Add possible moves to board
        for possible_move in moves {
            let index = self.square_index(&possible_move.square);
            formatted_board[index] = possible_move.value.to_string().yellow();
            println!("  {possible_move}");
        }
        // Print board with move positions
        let header: String = std::iter::once("    ".to_string())
            .chain(
                self.indices
                    .iter()
                    .map(|i| format!("{}", i.to_string().bold())),
            )
            .collect::<Vec<_>>()
            .join(" ");

        let board: String = self
            .indices
            .iter()
            .map(|y| {
                let row = self
                    .indices
                    .iter()
                    .map(|x| formatted_board[y * self.size + x].to_string())
                    .collect::<Vec<_>>()
                    .join(" ");
                format!("  {} {}", y.to_string().bold(), row)
            })
            .collect::<Vec<_>>()
            .join("\n");

        println!("{header}\n{board}");
    }

    /// Print current score for both players.
    pub fn print_score(&self) {
        let (black, white) = self.player_scores();
        println!("\n{self}");
        println!(
            "Score: {} | {}",
            black.to_string().magenta(),
            white.to_string().cyan()
        );
    }

    /// Returns the winning disk colour. Empty indicates a draw.
    pub fn result(&self) -> Disk {
        let total_score = self.score();
        match total_score.cmp(&0) {
            Ordering::Greater => Disk::White,
            Ordering::Less => Disk::Black,
            Ordering::Equal => Disk::Empty,
        }
    }

    /// Get board status string for game log.
    pub fn log_entry(&self) -> String {
        self.board.iter().map(|&disk| disk.board_char()).collect()
    }

    /// Check that the given coordinates are valid (inside the board).
    const fn check_coordinates(&self, x: isize, y: isize) -> bool {
        x >= 0 && x < self.size as isize && y >= 0 && y < self.size as isize
    }

    /// Check that the given square is valid (inside the board).
    const fn check_square(&self, square: &Square) -> bool {
        self.check_coordinates(square.x, square.y)
    }

    /// Returns the state of the board (empty, white, black) at the given square.
    fn get_square(&self, square: &Square) -> Option<Disk> {
        if self.check_square(square) {
            let index = self.square_index(square);
            Some(self.board[index])
        } else {
            None
        }
    }

    /// Map square to board index.
    const fn square_index(&self, square: &Square) -> usize {
        square.y as usize * self.size + square.x as usize
    }

    /// Count and return the number of black and white disks.
    fn player_scores(&self) -> (usize, usize) {
        // Alternative that is more readable but iterates the board twice:
        // let black = self.board.iter().filter(|&&disk| disk == Disk::Black).count();
        // let white = self.board.iter().filter(|&&disk| disk == Disk::White).count();
        let mut black = 0;
        let mut white = 0;
        for disk in &self.board {
            match disk {
                Disk::Black => black += 1,
                Disk::White => white += 1,
                Disk::Empty => {}
            }
        }
        (black, white)
    }

    /// Returns the total score.
    /// Positive value means more white disks and negative means more black disks.
    fn score(&self) -> i32 {
        self.board
            .iter()
            .filter(|&&disk| disk != Disk::Empty)
            .map(|disk| *disk as i32)
            .sum()
    }

    /// Sets the given square to the given value.
    fn set_square(&mut self, square: &Square, disk: Disk) {
        assert!(self.check_square(square), "Invalid coordinates: {square}");
        let index = self.square_index(square);
        self.board[index] = disk;
    }

    /// Initialize game board with starting disk positions.
    fn initialize_board(size: usize) -> Vec<Disk> {
        // Initialize game board with empty disks
        let mut board = vec![Disk::Empty; size * size];
        // Set starting positions
        let row = if size.is_multiple_of(2) {
            (size - 1) / 2
        } else {
            (size - 1) / 2 - 1
        };
        let col = size / 2;
        board[row * size + row] = Disk::White;
        board[row * size + col] = Disk::Black;
        board[col * size + row] = Disk::Black;
        board[col * size + col] = Disk::White;
        board
    }

    /// Initialize a set of empty squares for the board.
    fn initialize_empty_squares(size: usize, board: &[Disk]) -> HashSet<Square> {
        (0..board.len())
            .filter_map(|i| {
                if board[i] == Disk::Empty {
                    Some(Square {
                        x: (i % size) as isize,
                        y: (i / size) as isize,
                    })
                } else {
                    None
                }
            })
            .collect()
    }
}

impl fmt::Display for Board {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        // Horizontal indices
        write!(
            f,
            "  {}",
            self.indices
                .iter()
                .map(|&i| i.to_string())
                .collect::<Vec<String>>()
                .join(" ")
                .bold()
        )?;

        for y in &self.indices {
            // Vertical index
            write!(f, "\n{}", y.to_string().bold())?;
            // Row values
            for x in &self.indices {
                write!(
                    f,
                    " {}",
                    self.board[y * self.size + x].board_char_with_color()
                )?;
            }
        }

        Ok(())
    }
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
mod tests {
    use super::*;

    #[test]
    fn board_initialization() {
        let board_size = 8;
        let board = Board::new(board_size);
        assert_eq!(board.size, board_size);
        // Check initial disk positions
        assert_eq!(
            board.get_square(&Square { x: 3, y: 3 }).unwrap(),
            Disk::White
        );
        assert_eq!(
            board.get_square(&Square { x: 4, y: 4 }).unwrap(),
            Disk::White
        );
        assert_eq!(
            board.get_square(&Square { x: 3, y: 4 }).unwrap(),
            Disk::Black
        );
        assert_eq!(
            board.get_square(&Square { x: 4, y: 3 }).unwrap(),
            Disk::Black
        );
        // Rest of the board should be empty
        assert_eq!(
            board.get_square(&Square { x: 0, y: 0 }).unwrap(),
            Disk::Empty
        );
    }

    #[test]
    fn can_play() {
        let board = Board::new(8);
        assert!(board.can_play());
        // TODO: Fill the board and test again
    }

    #[test]
    fn scoring() {
        let board = Board::new(8);
        let (black_score, white_score) = board.player_scores();
        assert_eq!(black_score, 2);
        assert_eq!(white_score, 2);
        assert_eq!(board.score(), 0);
    }

    #[test]
    fn game_result() {
        let mut board = Board::new(8);
        let mut result = board.result();
        assert_eq!(result, Disk::Empty);

        let moves = board.possible_moves(Disk::Black);
        board.place_disk(&moves[0]);

        result = board.result();
        assert_eq!(result, Disk::Black);
    }

    #[test]
    fn log_entry() {
        let board = Board::new(8);
        let log_entry = board.log_entry();
        assert_eq!(
            log_entry,
            "___________________________WB______BW___________________________"
        );

        let mut board = Board::new(4);
        let log_entry = board.log_entry();
        assert_eq!(log_entry, "_____WB__BW_____");

        let moves = board.possible_moves(Disk::Black);
        board.place_disk(&moves[0]);
        let log_entry = board.log_entry();
        assert_eq!(log_entry, "____BBB__BW_____");
    }
}

File: typescript/src/board.ts
//==========================================================
// Board
// Defines the game board
// Akseli Lukkarila
// 2019-2026
//==========================================================

import chalk from 'chalk';

import { printYellow } from './colorprint.ts';
import { boardChar, boardCharWithColor, Direction, Disk, Move, opponent, Square, Step } from './models.ts';

const UP = 1;
const DOWN = -1;
const LEFT = -1;
const RIGHT = 1;
const STILL = 0;

/** All possible step directions for a square on the board. */
const STEP_DIRECTIONS: readonly Step[] = [
    new Step(DOWN, LEFT),
    new Step(DOWN, RIGHT),
    new Step(DOWN, STILL),
    new Step(STILL, LEFT),
    new Step(STILL, RIGHT),
    new Step(UP, LEFT),
    new Step(UP, RIGHT),
    new Step(UP, STILL),
];

/** Handles game board state and logic. */
export class Board {
    private board: Disk[];
    // Keep track of empty squares on board to avoid checking already filled positions.
    // Stores board indices since JS Set compares objects by reference.
    private emptySquares: Set<number>;
    // Index list (0...size) to avoid repeating same range in loops.
    private readonly indices: number[];
    readonly size: number;

    /** Initialize a new board for the given board size. */
    constructor(size: number) {
        this.board = Board.initBoard(size);
        this.emptySquares = Board.initEmptySquares(this.board);
        this.indices = Array.from({ length: size }, (_, i) => i);
        this.size = size;
    }

    /** Return true if board contains empty squares. */
    canPlay(): boolean {
        return this.emptySquares.size > 0;
    }

    /** Update board for given disk placement. */
    placeDisk(chosenMove: Move): void {
        const start = chosenMove.square;
        const square = this.getSquare(start);
        if (square === undefined) {
            throw new Error(`Invalid coordinates: ${start}`);
        }
        if (square !== Disk.Empty) {
            throw new Error(`Trying to place disk to an occupied square: ${start}!`);
        }
        this.setSquare(start, chosenMove.disk);
        this.emptySquares.delete(this.squareIndex(start));
        for (const affectedSquare of chosenMove.affectedSquares()) {
            this.setSquare(affectedSquare, chosenMove.disk);
        }
    }

    /** Returns a list of possible moves for the given player. */
    possibleMoves(disk: Disk): Move[] {
        const moves: Move[] = [];
        const opposingDisk = opponent(disk);
        for (const index of this.emptySquares) {
            const square = new Square(index % this.size, Math.floor(index / this.size));
            let value = 0;
            const directions: Direction[] = [];
            for (const step of STEP_DIRECTIONS) {
                let pos = square.add(step);
                // Next square in this direction needs to be the opposing disk
                if ((this.getSquare(pos) ?? Disk.Empty) !== opposingDisk) {
                    continue;
                }
                let numSteps = 0;
                // Keep stepping over opponents disks
                while ((this.getSquare(pos) ?? Disk.Empty) === opposingDisk) {
                    numSteps += 1;
                    pos = pos.add(step);
                }
                // Valid move only if a line of opposing disks ends with own disk
                if ((this.getSquare(pos) ?? Disk.Empty) === disk) {
                    directions.push(new Direction(step, numSteps));
                    value += numSteps;
                }
            }
            if (value > 0) {
                moves.push(new Move(square, disk, value, directions));
            }
        }
        moves.sort(Move.compare);
        return moves;
    }

    /** Print board with available move coordinates and the resulting points gained. */
    printPossibleMoves(moves: Move[]): void {
        printYellow(`  Possible moves (${moves.length}):`);
        // Convert board from Disk enums to strings
        const formattedBoard = this.board.map(boardCharWithColor);
        // Add possible moves to board
        for (const possibleMove of moves) {
            const index = this.squareIndex(possibleMove.square);
            formattedBoard[index] = chalk.yellow(possibleMove.value.toString());
            console.log(`  ${possibleMove}`);
        }
        // Print board with move positions
        let text = '    ';
        for (const i of this.indices) {
            text += ` ${chalk.bold(i.toString())}`;
        }
        for (const y of this.indices) {
            text += `\n  ${chalk.bold(y.toString())}`;
            for (const x of this.indices) {
                text += ` ${formattedBoard[y * this.size + x]}`;
            }
        }
        console.log(text);
    }

    /** Print current score for both players. */
    printScore(): void {
        const [black, white] = this.playerScores();
        console.log(`\n${this}`);
        console.log(`Score: ${chalk.magenta(black.toString())} | ${chalk.cyan(white.toString())}`);
    }

    /** Returns the winning disk colour. Empty indicates a draw. */
    result(): Disk {
        const totalScore = this.score();
        if (totalScore === 0) {
            return Disk.Empty;
        }
        return totalScore > 0 ? Disk.White : Disk.Black;
    }

    /** Get board status string for game log. */
    logEntry(): string {
        return this.board.map(boardChar).join('');
    }

    /** Check that the given coordinates are valid (inside the board). */
    private checkCoordinates(x: number, y: number): boolean {
        return x >= 0 && x < this.size && y >= 0 && y < this.size;
    }

    /** Check that the given square is valid (inside the board). */
    private checkSquare(square: Square): boolean {
        return this.checkCoordinates(square.x, square.y);
    }

    /** Returns the state of the board (empty, white, black) at the given square. */
    private getSquare(square: Square): Disk | undefined {
        return this.checkSquare(square) ? this.board[this.squareIndex(square)] : undefined;
    }

    /** Map square to board index. */
    private squareIndex(square: Square): number {
        return square.y * this.size + square.x;
    }

    /** Count and return the number of black and white disks. */
    private playerScores(): [number, number] {
        let black = 0;
        let white = 0;
        for (const disk of this.board) {
            switch (disk) {
                case Disk.Black:
                    black += 1;
                    break;
                case Disk.White:
                    white += 1;
                    break;
                default:
                    break;
            }
        }
        return [black, white];
    }

    /**
     * Returns the total score.
     * Positive value means more white disks and negative means more black disks.
     */
    private score(): number {
        return this.board.reduce((sum, disk) => sum + disk, 0);
    }

    /** Sets the given square to the given value. */
    private setSquare(square: Square, disk: Disk): void {
        if (!this.checkSquare(square)) {
            throw new Error(`Invalid coordinates: ${square}`);
        }
        this.board[this.squareIndex(square)] = disk;
    }

    /** Initialize game board with starting disk positions. */
    private static initBoard(size: number): Disk[] {
        // Initialize game board with empty disks
        const board: Disk[] = new Array<Disk>(size * size).fill(Disk.Empty);
        // Set starting positions
        const half = Math.floor((size - 1) / 2);
        const row = size % 2 === 0 ? half : half - 1;
        const col = Math.floor(size / 2);
        board[row * size + row] = Disk.White;
        board[row * size + col] = Disk.Black;
        board[col * size + row] = Disk.Black;
        board[col * size + col] = Disk.White;
        return board;
    }

    /** Initialize empty squares for the board. */
    private static initEmptySquares(board: Disk[]): Set<number> {
        const emptySquares = new Set<number>();
        for (let index = 0; index < board.length; index++) {
            if (board[index] === Disk.Empty) {
                emptySquares.add(index);
            }
        }
        return emptySquares;
    }

    /** Format game board to string. */
    toString(): string {
        // Horizontal indices
        let text = `  ${chalk.bold(this.indices.join(' '))}`;
        for (const y of this.indices) {
            // Vertical index
            text += `\n${chalk.bold(y.toString())}`;
            // Row values
            for (const x of this.indices) {
                text += ` ${boardCharWithColor(this.board[y * this.size + x])}`;
            }
        }
        return text;
    }
}
